diff --git a/swift/test/Misc/verify-swift-feature-testing.test-sh b/swift/test/Misc/verify-swift-feature-testing.test-sh
index 3f36fc979cb..8713ae8fea8 100755
--- a/swift/test/Misc/verify-swift-feature-testing.test-sh
+++ b/swift/test/Misc/verify-swift-feature-testing.test-sh
@@ -98,7 +98,6 @@ def find_matches(swift_src_root):
     # `-enable-upcoming-feature` in the test directories.
     output = subprocess.check_output(
         [
-            "git",
             "grep",
             "--extended-regexp",
             "--recursive",
diff --git a/swift/test/SILOptimizer/dead_array_elim.swift b/swift/test/SILOptimizer/dead_array_elim.swift
index a15d6aa2254..8426e66c7e0 100644
--- a/swift/test/SILOptimizer/dead_array_elim.swift
+++ b/swift/test/SILOptimizer/dead_array_elim.swift
@@ -2,7 +2,6 @@
 
 // REQUIRES: swift_stdlib_no_asserts
 // REQUIRES: swift_in_compiler
-// XFAIL: OS=linux-androideabi
 
 // These tests check whether DeadObjectElimination pass runs as a part of the
 // optimization pipeline and eliminates dead array literals in Swift code.
